### PR TITLE
Support Java overloaded instance methods

### PR DESCRIPTION
**Pull request by jponge**
This supports resolving and invalidating Java overloaded instance methods.

### CODE REVIEW

File: src/main/java/org/eclipse/golo/runtime/RegularMethodFinder.java [hunk 1]
```diff
@@ -9,44 +9,48 @@
...
+    return Extractors.getMethods(invocation.receiverClass())
+        .filter(this::overloadMatch)
+        .count() > 1;
+  }
+
+  private boolean overloadMatch(Method m) {
+    return Extractors.isPublic(m) &&
+        Extractors.isConcrete(m) &&
+        m.getName().equals(invocation.name()) &&
+        (m.getParameterCount() + 1 == invocation.arity()) || (m.isVarArgs() && (m.getParameterCount() <= invocation.arity()));
```
Comment (yloiseau) [inline src/main/java/org/eclipse/golo/runtime/RegularMethodFinder.java]: looks like déjà vu... maybe `TypeMatching.argumentsNumberMatches` ?

File: src/main/java/org/eclipse/golo/runtime/RegularMethodFinder.java [hunk 2]
```diff
@@ -82,37 +86,39 @@ public boolean isOverloaded() {
...
   }
 
   private static String reverseStructAugmentation(String receiverClassName) {
     return receiverClassName.substring(0, receiverClassName.indexOf(".types"))
-      + "$" + receiverClassName.replace('.', '$');
+        + "$" + receiverClassName.replace('.', '$');
   }
 
 
   private Stream<Method> findInMethods() {
```
Comment (danielpetisme) [inline src/main/java/org/eclipse/golo/runtime/RegularMethodFinder.java]: Is it meaningful to extract `findInMethods`, `findInFields`, `toMethodHandles` and others into a Reflect/InvokeUtils ?

Comment (yloiseau) [inline src/main/java/org/eclipse/golo/runtime/RegularMethodFinder.java]: We indeed have a code duplication (e.g. in `FunctionCallSupport` and `Predefined::fun`), but AFAIR, these particular methods are specific to a particular finder, so I don't think we can really extract common code. Type matching, checking and filtering are also candidates to possible duplication (even if I tried to remove as much as possible).
BTW, I started to extract the naming logic into a separate utility class, but didn't integrated it in my previous refactoring PR. Indeed, we are doing special names manipulations (like `MyModule.types.MyStruct`, augmentations replacing `.` with `$`, etc.) in several separated places: IR construction, bytecode generation, runtime lookup to name a few.

File: src/main/java/org/eclipse/golo/runtime/RegularMethodFinder.java (hunk 1 above)
Comment (jponge) [inline src/main/java/org/eclipse/golo/runtime/RegularMethodFinder.java]: I need to try it :-)

Comment (jponge) [inline src/main/java/org/eclipse/golo/runtime/RegularMethodFinder.java]: No, it doesn't fit.

**Comment by jponge** [comment]
@yloiseau This makes me think that `FunctionCallSupport` is a good refactoring candidate.

This PR focuses only on instance methods, but once refactored a bit we could improve `FunctionCallSupport` to also deal with overloaded Java static methods.

**Comment by jponge** [comment]
Recorded #341 for overloaded static methods.

I'm ok with this PR, if you agree then I'll go ahead and merge.

**Comment by yloiseau** [comment]
:+1:

**Pull request merged**